cpp/src/Ice/ThreadPool.cpp

```diff
@@ -556,37 +510,26 @@ IceInternal::ThreadPool::executeFromThisThread(const ExecutorWorkItemPtr& workIt
     }
     else
     {
-        workItem->run();
+        call();
     }
 #endif
 }
 
 void
-IceInternal::ThreadPool::execute(const ExecutorWorkItemPtr& workItem)
+IceInternal::ThreadPool::execute(function<void()> call, const Ice::ConnectionPtr& connection)
 {
     lock_guard lock(_mutex);
     if (_destroyed)
     {
         throw CommunicatorDestroyedException(__FILE__, __LINE__);
     }
-    _workQueue->queue(workItem);
-}
-
-void
-IceInternal::ThreadPool::execute(function<void()> call)
-{
-    class WorkItem final : public IceInternal::ExecutorWorkItem
-    {
-    public:
-        WorkItem(function<void()> call) : _call(std::move(call)) {}
 
-        void run() final { _call(); }
-
-    private:
-        function<void()> _call;
-    };
-
-    execute(make_shared<WorkItem>(std::move(call)));
+    _workQueue->queue(
+        [self = shared_from_this(), call = std::move(call), connection](ThreadPoolCurrent& current)
```

Comment:
It's not clear to me what the type of connection is here.

Maybe it would be cleaner for the execute's connection parameter to be a copy (`Ice::ConnectionPtr`) to make clear we copy it?
In practice, `call` (always?) holds onto the connection so it won't change anything.

Reply:
I'm afraid I don't understand your proposal. Can you perhaps post a code sample?

Reply:
I was confused. `[..., connection]` is captured by value here, meaning the capture holds onto a new `ConnectionPtr`.